core/modules/script.py

```diff
@@ -0,0 +1,154 @@
+            path += data["name"]
+        else:
+            if skel and "name" in skel:
+                path += skel["name"]
+    elif skel:
+        if skel["name"]:
+            path += skel["name"]
+
+    return path
+
+
+class BaseScriptAbstractSkel(TreeSkel):
+    PATH_SUFFIX = ""
+
+    path = StringBone(
+        descr="Path",
+        required=True,
+        unique=UniqueValue(UniqueLockMethod.SameValue, True, "The entered file name already exists.")
+    )
+
+    @classmethod
+    def set_path(cls, skel, data):
+        skel["path"] = utils_get_path(skel, data) + cls.PATH_SUFFIX
+
+        if data:
+            data["path"] = skel["path"]
+
+    @classmethod
+    def fromClient(cls, skel, data, *args, **kwargs):
+        cls.set_path(skel, data)
+
+        ret = super().fromClient(skel, data, *args, **kwargs)
+
+        if not ret:
+            # in case the identifier failed because the unique value is already provided, rewrite the error for variable
+            for error in skel.errors:
+                if error.severity == skeleton.ReadFromClientErrorSeverity.Invalid and error.fieldPath == ["path"]:
+                    error.fieldPath = ["path"]
+                    break
+
+        return ret
+
+
+class ScriptNodeSkel(BaseScriptAbstractSkel):
+    PATH_SUFFIX = "/"
+    kindName = "viur-script-node"
+
+    rootNode = BooleanBone(
+        descr="Is root node?",
+        defaultValue=False,
+    )
+
+    plugin = BooleanBone(
+        descr="Is plugin?",
+        defaultValue=False
+    )
+
+    name = StringBone(
+        descr="Folder",
+        required=True,
+        vfunc=lambda value: not DIRECTORY_PATTERN.match(value)
+    )
+
+
+class ScriptLeafSkel(BaseScriptAbstractSkel):
+    kindName = "viur-script-leaf"
+
+    name = StringBone(
+        descr="Filename",
+        vfunc=lambda value: not FILE_PATTERN.match(value),
+    )
+
+    script = RawBone(
+        descr="Code",
+        indexed=False
+    )
+
+
+class Script(Tree):
+    leafSkelCls = ScriptLeafSkel
+    nodeSkelCls = ScriptNodeSkel
+
+    def adminInfo(self):
+        return conf.get("viur.script.admin_info") or {}
+
+    def getAvailableRootNodes(self):
+        if not utils.getCurrentUser():
+            return []
+
+        return [{"name": "Scripts", "key": self.ensureOwnModuleRootNode().key}]
+
+    def onEdit(self, skelType, skel):
+        path = utils_get_path(skel)
+        if skelType == "node":
+            path += ScriptNodeSkel.PATH_SUFFIX
+
+        skel["path"] = path
+
+    def onEdited(self, skelType, skel):
+        if skelType == "node":
+            query = self.editSkel("node").all().filter("parententry =", skel["key"])
+
+            for entry in query.fetch(99):
+                self.onEdit("node", entry)
+                entry.toDB()
+
+            query = self.editSkel("leaf").all().filter("parententry =", skel["key"])
+            for entry in query.fetch(99):
+                self.onEdit("leaf", entry)
+                entry.toDB()
```

Comment:
I'm not sure if we really should run this in the same request. This could lead to a long response time which might not be user-friendly.

Reply:
Then we should optimize this previously.

Reply:
I've optimized this.